File: scripts/compile_meson.js
const { readdirSync, statSync, writeFileSync, existsSync } = require('fs');
const { extname } = require('path');

/**
 * 
 * @param {string} directory 
 */
function readDirectoryRecursive(directory) {

    if (!existsSync(directory)) throw new Error(`Directory does not exist: ${directory}`);

    const root = directory
    const files = readdirSync(root);

    let built = `# root/${root}meson.build\n\n`

    for (let file of files) {

        const path = root + file;
        const _stat = statSync(path);

        if (_stat.isDirectory()) {
            built += `subdir('${file}')\n`;
            readDirectoryRecursive(path + '/');
            continue;
        }

        if (extname(path).toLowerCase() !== '.cpp') continue;

        if (file.toLowerCase().includes('main.cpp')) {
            built += `src_files = files('${file}')\n`
        }
        else built += `src_files += files('${file}')\n`

    }

    // check if meson.build already exists
    if (existsSync(root + 'meson.build') && statSync(root + 'meson.build').isFile()) {
        var existing = require('fs').readFileSync(root + 'meson.build', 'utf-8');
        if (existing === built) {
            console.log(`${root}meson.build... Skipped (up to date).`);
            return;
        }
        else {
            console.log(`${root}meson.build... Updated.`);
        }
    }
    else {
        console.log(`${root}meson.build... Created.`);
    }

    writeFileSync(`${root}meson.build`, built);
}

readDirectoryRecursive('src/')
console.log('OK.');

File: scripts/compile_meson.py
import sys
from pathlib import Path

def readDirRecursive(directory: Path):
    if not directory.exists():
        print(f"Directory '{directory}' does not exist.")
        sys.exit()

    root = directory
    built = f"# root/{root.as_posix()}/meson.build\n\n"

    subdirs = []
    cpp_files = []

    for file in Path(root).glob("*"):
        path: Path = root / file.name
        if file.is_dir():
            subdirs.append(file.name)
            readDirRecursive(path)
        elif file.name.endswith(".cpp"):
            cpp_files.append(file.name)

    for name in cpp_files:
        if name.lower() == "main.cpp":
            built += f"src_files = files('{name}')\n"
        else:
            built += f"src_files += files('{name}')\n"

    for name in subdirs:
        built += f"subdir('{name}')\n"

    build_file = Path(root / "meson.build")
    if build_file.exists() and build_file.is_file():
        build_file_content = build_file.read_text()
        if build_file_content != built:
            build_file.write_text(built)
            print(f"meson.build... Updated.")
    else:
        build_file.write_text(built)
        print(f"meson.build... Created.")

if len(sys.argv) != 2:
    print("Usage: python compile_meson.py <directory>")
    sys.exit()

readDirRecursive(Path(sys.argv[1]))
print("Meson OK.")